src/decoder.rs

```diff
@@ -89,10 +89,13 @@ impl Decoder {
                 self.buffer.push(byte);
                 if self.buffer.len() == 2 {
                     self.message.checksum = u16::from_le_bytes([self.buffer[0], self.buffer[1]]);
+                    self.reset();
                     let message = self.message.clone();
                     self.message = ProtocolMessage::new();
-                    self.reset();
-                    return DecoderResult::Success(message);
+                    if message.has_valid_crc() {
+                        return DecoderResult::Success(message);
+                    }
+                    return DecoderResult::Error(ParseError::ChecksumError(message));
```

Comment:
Just to keep it clearer, keeping the main flow not in the some branch

```suggestion
                    if !message.has_valid_crc() {
                        return DecoderResult::Error(ParseError::ChecksumError(message));                        
                    }

                    return DecoderResult::Success(message);
```